Feature/pace extrapolation kokkos

yury-lysogorskiy opened the pull request:
**Summary**

Implementation of extrapolation grade gamma (CPU implementation is here: https://github.com/lammps/lammps/pull/3315)  with KOKKOS

**Related Issue(s)**

<!--If this addresses an open GitHub issue for this project, please mention the issue number here, and describe the relation. Use the phrases `fixes #221` or `closes #135`, when you want an issue to be automatically closed when the pull request is merged-->

**Author(s)**

Yury Lysogorskiy (ICAMS, Ruhr-University Bochum)

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**


**Implementation Notes**

* new KOKKOS/pair_pace_extrapolation_kokkos.h/cpp  files, copy of  KOKKOS/pair_pace_kokkos.h/cpp
* modified kernel `TagPairPACEComputeRho` to mimic ACEBEvaluator ([here](https://github.com/ICAMS/lammps-user-pace/blob/main/ML-PACE/ace/ace_b_evaluator.cpp#L332) and  [here](https://github.com/ICAMS/lammps-user-pace/blob/main/ML-PACE/ace/ace_b_evaluator.cpp#L397)) and also to collect B-projections
* new kernel `TagPairPACEComputeGamma`

**Post Submission Checklist**

<!--Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply-->

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

<!--Put any additional information here, attach relevant text or image files, and URLs to external sites (e.g. DOIs or webpages)-->

yury-lysogorskiy committed:
WIP: add pair_pace_extrapolation_kokkos.cpp/h

yury-lysogorskiy committed:
WIP: pair_pace_extrapolation_kokkos.h/cpp:
- rename idx_rho_max -> idx_ms_combs_max,  d_idx_rho_count ->d_idx_ms_combs_count, d_offsets->d_func_inds
- remove d_ctildes, add d_gen_cgs and d_coeffs
- use ACEBBasisFunction
- update TagPairPACEComputeRho and TagPairPACEComputeWeights
- pair_pace_extrapolation.h/cpp: add chunksize option

yury-lysogorskiy committed:
WIP:
- add TagPairPACEComputeGamma kernel
- add d_total_basis_size
- add gamma_flag, d_ASI, projections and gamma

yury-lysogorskiy committed:
- add extract and extract_peratom methods
- rename device gamma array to d_gamma
- make host h_gamma array
- copy from h_gamma to host extrapolation_grade_gamma array for each chunk
- transpose two last dimensions of d_ASI (small improvement of performance)
- manage grows of extrapolation_grade_gamma

yury-lysogorskiy committed:
update lammps-user-pace version and checksum

yury-lysogorskiy committed:
add check that LINEAR ASI must be used

yury-lysogorskiy committed:
update pair_pace.rst

akohlmey commented on src/KOKKOS/pair_pace_extrapolation_kokkos.h:
```diff
@@ -0,0 +1,354 @@
+#define LMP_PAIR_PACE_EXTRAPOLATION_KOKKOS_H
+
+#include "pair_pace_extrapolation.h"
+#include "ace-evaluator/ace_radial.h"
```
This include should be avoided. Any header declaring a "style" should not include any implementation specific headers.

akohlmey replied:
Upon closer inspection, this change **must** be done. The include file in question is in violation of the programming style requirement for LAMMPS. The standout issue is the use of "using namespace std;" in a header file. This effectively eliminates the benefits of namespaces and because all style files of the same kind are included at the same time in LAMMPS, will be implicitly applied to many other headers and classes and thus potentially cause namespace conflicts or unexpected access ambiguities beyond the style where it is applied. We've had very hard to debug problems of this kind before where adding a package would unexpectedly break some other packages.

In most of LAMMPS we go one step further and do not even use "using namespace std;" in the implementation to make it more obvious to people reading the code which are functions/classes of the C++ standard library and which are functions/members/instances of LAMMPS classes.

yury-lysogorskiy committed:
fix_pair.cpp: respect lmp->suffix when looking for pair_style name match

stanmoore1 committed:
Small cleanup

stanmoore1 commented on src/fix_pair.cpp:
```diff
@@ -36,6 +36,10 @@ FixPair::FixPair(LAMMPS *lmp, int narg, char **arg) :
   if (nevery < 1) error->all(FLERR,"Illegal fix pair every value: {}", nevery);
 
   pairname = utils::strdup(arg[4]);
+  if(lmp->suffix) {
```
In other places we use the `^` regex. This logic will not work for `suffix2`.

stanmoore1 replied:
@akohlmey what do you suggest for this?

stanmoore1 replied:
@akohlmey thanks, seems like it may be better to have a `force->pair_match_suffix` that pulls in all the suffix logic?

akohlmey replied:
That would be a larger refactoring project. The challenge is to update all use cases and be mindful of exceptions.
The current API has grown over time. There is suffix handling and the look up of hybrid sub-styles. Both were added incrementally and I noticed that not all applications do both and do it consistently.

stanmoore1 commented:
@yury-lysogorskiy nice job, thanks for working on this!

akohlmey committed:
update / correct suffix handling in fix pair

akohlmey reviewed (changes requested):
None of the "style" headers or any of the headers they include may use "using namespace std;"
If at all possible the implementation should be hidden from the interface and forward declarations or a pointer to a struct with the actual implementation stored (PIMPL). That way the headers for a style only need to include LAMMPS headers.

yury-lysogorskiy committed:
- replace #include "ace-evaluator/ace_radial.h" in pair_pace_kokkos.h and pair_pace_extrapolation_kokkos.h with forward declaration "class SplineInterpolator;"
- move SplineInterpolatorKokkos::operator=(const SplineInterpolator &spline) to .cpp files

yury-lysogorskiy commented:
`fix pair` doesn't work anymore,  trying to fix it, please, do not merge

yury-lysogorskiy committed:
refactor fix_pair.h/cpp: extract method "query_pstyle" and call it also in void FixPair::init()

yury-lysogorskiy commented:
I fixed the issue with `fix pair` in [20b6355](https://github.com/lammps/lammps/pull/3585/commits/20b6355888d3b005601b62439270cd749dae24e9)

akohlmey commented:
@stanmoore1 this is still assigned to you. Are you still testing?

stanmoore1 commented:
@akohlmey I approve the code changes and believe it is ready to be merged.
